src/primitives/transaction.cpp

```diff
@@ -9,84 +9,56 @@
+            const boost::array<libzcash::JSOutput, ZC_NUM_JS_OUTPUTS>& outputs,
             CAmount vpub_old,
-            CAmount vpub_new) : scriptSig(), scriptPubKey(scriptPubKey), vpub_old(vpub_old), vpub_new(vpub_new), anchor(anchor)
+            CAmount vpub_new) : vpub_old(vpub_old), vpub_new(vpub_new), anchor(anchor)
 {
-    uint256 scriptPubKeyHash;
-    {
-        CHashWriter ss(SER_GETHASH, 0);
-        ss << scriptPubKey;
-        scriptPubKeyHash = ss.GetHash();
-    }
-
-    PourTransaction pourtx(params,
-                           std::vector<unsigned char>(scriptPubKeyHash.begin(), scriptPubKeyHash.end()),
-                           std::vector<unsigned char>(anchor.begin(), anchor.end()),
-                           std::vector<PourInput>(inputs.begin(), inputs.end()),
-                           std::vector<PourOutput>(outputs.begin(), outputs.end()),
-                           vpub_old,
-                           vpub_new);
-
-    boost::array<std::vector<unsigned char>, NUM_POUR_INPUTS> serials_bv;
-    boost::array<std::vector<unsigned char>, NUM_POUR_OUTPUTS> commitments_bv;
-    boost::array<std::vector<unsigned char>, NUM_POUR_INPUTS> macs_bv;
-
-    proof = pourtx.unpack(serials_bv, commitments_bv, macs_bv, ciphertexts, ephemeralKey);
-    serials = unsigned_char_vector_array_to_uint256_array(serials_bv);
-    commitments = unsigned_char_vector_array_to_uint256_array(commitments_bv);
-    macs = unsigned_char_vector_array_to_uint256_array(macs_bv);
+    boost::array<libzcash::Note, ZC_NUM_JS_OUTPUTS> notes;
+
+    params.loadProvingKey();
+    proof = params.prove(
+        inputs,
+        outputs,
+        notes,
+        ciphertexts,
+        ephemeralKey,
+        pubKeyHash,
+        randomSeed,
+        macs,
+        serials,
+        commitments,
+        vpub_old,
+        vpub_new,
+        anchor
+    );
 }
 
-bool CPourTx::Verify(ZerocashParams& params) const {
-    // Compute the hash of the scriptPubKey.
-    uint256 scriptPubKeyHash;
-    {
-        CHashWriter ss(SER_GETHASH, 0);
-        ss << scriptPubKey;
-        scriptPubKeyHash = ss.GetHash();
-    }
-
-    return PourProver::VerifyProof(
-        params,
-        std::vector<unsigned char>(scriptPubKeyHash.begin(), scriptPubKeyHash.end()),
-        std::vector<unsigned char>(anchor.begin(), anchor.end()),
+bool CPourTx::Verify(
+    ZCJoinSplit& params,
+    const uint256& pubKeyHash
+) const {
+    return params.verify(
+        proof,
+        pubKeyHash,
+        randomSeed,
+        macs,
+        serials,
+        commitments,
         vpub_old,
         vpub_new,
-        uint256_to_array<NUM_POUR_INPUTS>(serials),
-        uint256_to_array<NUM_POUR_OUTPUTS>(commitments),
-        uint256_to_array<NUM_POUR_INPUTS>(macs),
-        proof
+        anchor
     );
 }
 
+uint256 CPourTx::h_sig(ZCJoinSplit& params, const uint256& pubKeyHash) const
+{
+    return params.h_sig(randomSeed, serials, pubKeyHash);
```

Comment:
Yes, it's right, assuming `h_sig` takes its arguments in the same order as the spec.